pkg/restclient/roundtrippers.go

```diff
@@ -0,0 +1,88 @@
+package restclient
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// debugRoundTripper is a middleware which debugs each outgoing request.
+// It will dump information about request passing through it.
+type debugRoundTripper struct {
+	rt http.RoundTripper
+}
+
+func newDebugRoundTripper(rt http.RoundTripper) http.RoundTripper {
+	return &debugRoundTripper{rt}
+}
+
+// RoundTrip dumps the request and response of a single http transaction.
+//
+// It will dump each outgoing request in its HTTP/1.x wire representation and
+// also It will dump the received response in its HTTP/1.x wire representation.
+func (d *debugRoundTripper) RoundTrip(req *http.Request) (resp *http.Response, err error) {
+	reqCtx := newRequestCtx(req)
+	logrus.Debug(reqCtx.curl())
+
+	if reqDump, err := httputil.DumpRequestOut(req, true); err == nil {
+		logrus.Debugf("Trying %s... \n%s", reqCtx.reqURL.Host, reqCtx.dump(">", reqDump))
+	}
+
+	defer func(begin time.Time) {
+		respDur := time.Since(begin)
+		if err != nil {
+			logrus.Debugf("gd2client error: %s", err)
+			return
+		}
+
+		logrus.Debugf("%s %s %s in %s", reqCtx.reqVerb, reqCtx.reqURL, resp.Status, respDur.String())
+		if respDump, err := httputil.DumpResponse(resp, true); err == nil {
+			logrus.Debug("\n", reqCtx.dump("<", respDump))
+		}
+	}(time.Now())
+
+	return d.rt.RoundTrip(req)
+}
+
+type requestCtx struct {
+	reqVerb   string
+	reqURL    *url.URL
+	reqHeader http.Header
+}
+
+func newRequestCtx(req *http.Request) *requestCtx {
+	r := &requestCtx{
+		reqHeader: make(http.Header),
+		reqURL:    new(url.URL),
+	}
+	r.reqVerb = req.Method
+	*r.reqURL = *req.URL
+	r.reqHeader = req.Header
+	return r
+}
+
+func (r *requestCtx) curl() string {
+	var header string
+	for key, vals := range r.reqHeader {
+		for _, val := range vals {
+			header += fmt.Sprintf(` -H %q`, fmt.Sprintf("%s : %s", key, val))
+		}
+	}
+	return fmt.Sprintf("curl -v -k -X%s %s '%s'", r.reqVerb, header, r.reqURL)
+}
+
+func (r *requestCtx) dump(prefix string, data []byte) string {
```

Comment:
This needn't be a method on `requestCtx`. `r *requestCtx` isn't being used anywhere within.